docs/graphql.md

```diff
@@ -0,0 +1,141 @@
+
+```graphql
+query UserWithComments {
+  user {
+    id
+    name
+    comments {
+      body
+    }
+  }
+}
+```
+
+The point of these are to provide an easy way for the developers, when things go wrong, to pinpoint where exactly they did so.
+
+This name is available on the server too and serves as a great distinguishing key.
+
+Span name examples:
+- `GraphQL:UserWithComments`
+- `GraphQL:UpdateUser`
+
+## Multiple endpoints
+
+An app may serve multiple GraphQL endpoints. To tell them apart we can include the path in the transaction name.
+
+- `GraphQL:UserWithComments (/api/graphql)`
+
+This does not seem very common, so adding this is an opt-in option, like `graphql_postfix_path: true`.
+
+## Anonymous queries
+
+An Operation Name isn't required. When one isn't provided it's hard for us to tell apart the queries.
+
+1. Some clients generate `id`s from hashing the contents of the query (see [apollo-tooling](https://github.com/apollographql/apollo-tooling/blob/1dfd737eaf85b89b2cfb13913342e091e3c03d18/packages/apollo-codegen-core/src/compiler/visitors/generateOperationId.ts#L5)). This would split the anonymous queries into separate buckets.
+
+    A problem with this approach is that a user of the APM UI has no way to recognise queries in the transactions list before clicking through.
+
+    Using just the `id` will not reveal the true culprit since there can be variables associated with the query. Different values for the variables can lead to very different workloads and response times.
+
+2. Another approach is to simply label them `[anonymous]` or something similar.
+
+    A problem with _that_ approach is that the contents and thereby the relevant db queries and other sub-span actions that the server might do while resolving these queries may be wildly different making it hard to provide a _true_ sample waterfall.
+
+    These two examples for example will look the same for the top-level GraphQL spans but will represent significantly different workloads.
+
+    ```
+    [- anonymous graphql span --------------]
+      [- 1,000x SELECT * ---------------]
+        [- 1,000 more SELECT * -]
+
+    [- anonymous graphql span --------------]
+      [- SELECT id FROM users WHERE id=? -]
+    ```
+
+    We could consider _muting_ or ignoring all sub-spans to anonymous GraphQL queries and choose to rather show nothing than potentially wrong information.
+
+No one of these are perfect. Because the benefits of using `id`s in the worst case could be misleading anyway, we're going with option 2.
+
+To further help and nudge developers to use Operation Names for their queries, we could log a warning when coming across anonymous queries.
+
+```plain
+[DEBUG] Traced anonymous GraphQL query. Use Operation Names for your queries to better distinguish them in the APM UI.
+```
+
+This could also be a tooltip in the UI instead of a log message.
+
+Span name examples:
+- `GraphQL:ka8kadf8233kxcsc2929384384kdkdkc8383…`
+- `GraphQL:[anonymous]`
```

Comment:
How would the corresponding transaction name look? Would it also use "[anonymous]"?

Reply:
I guess? I still think being GraphQL is the most significant knowledge of the request. Alternatively it would fall back to `POST /graphql` which, next to named GraphQL queries, would probably be confusing.